File: 백준/Silver/1926. 그림/그림.cc
#include <iostream>
#include <vector>
#include <string>
#include <queue>
#include "algorithm"
using namespace std;

int board[501][501];
int vis[501][501];

int dx[4]{ 0, 0, -1, 1 };
int dy[4]{ 1, -1, 0, 0 };


int main(void)
{
    ios::sync_with_stdio(0);
    cin.tie(0);

    // 행과 열을 저장해줄 변수
    int row, col;

    cin >> row >> col;

    // Board 셋팅
    for (size_t i = 0; i < row; i++)
        for (size_t j = 0; j < col; j++)
            cin >> board[i][j];

    // 그림의 수
    int count{};

    // 찾은 그림들 중 가장 큰 영역의 수
    int mx{};

    // 전수 검사를 위해 2중 포문안에 BFS를 넣어주어야함
    for (size_t i = 0; i < row; i++)
    {
        for (size_t j = 0; j < col; j++)
        {
            if (vis[i][j] == 1 || board[i][j] == 0)
                continue;

            // 갈 수 있는 위치를 저장해주는 q
            queue<pair<int, int>> q;
            q.push({ i, j });

            // 방문
            vis[i][j] = 1;

            // 새로운 BFS가 시작되는 시점이 새로운 영역(그림)이라는 뜻
            ++count;

            // 그림마다 영역을 체크해줄 일회성 변수
            int area{};

            // BFS 시작
            while (!q.empty())
            {
                ++area;

                pair<int, int> cur = q.front();
                q.pop();

                for (size_t i = 0; i < 4; i++)
                {
                    int nx = cur.second + dx[i];
                    int ny = cur.first + dy[i];

                    if (nx < 0 || nx >= col || ny < 0 || ny >= row)
                        continue;
                    if (vis[ny][nx] == 1 || board[ny][nx] == 0)
                        continue;

                    vis[ny][nx] = 1;
                    q.push({ ny, nx });
                }
            }
            
            // 가장 큰 영역 교체
            mx = max(mx, area);
        }
    }

    cout << count << "\n" << mx;

    return 0;
}       

// 그림의 수와 그림들 중 가장 큰 영역의 수 출력하기 ->줄바꿈
// BFS 전수검사 문제
// 그림의 수를 세어줄 변수와 영역크기를 세어줄 변수가 필요함
// 행 열 순으로 받고, int 2차원 배열로 받는다
// vis가 필요한 문제

File: 백준/Silver/1926. 그림/그림.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Security.AccessControl;
using System.Text;
using System.Threading.Tasks;

class program
{
   
    static void Main()
    {
        int[] dx = new int[4] { 0, 0, -1, 1 };
        int[] dy = new int[4] { -1, 1, 0, 0 };

        var sr = new StreamReader(Console.OpenStandardInput());

        string[] s = sr.ReadLine().Split();
        int n = int.Parse(s[0]); 
        int m = int.Parse(s[1]); 

        int[,] board = new int[n,m];

        for (int i = 0; i < n; i++)
        {
            string[] input = sr.ReadLine().Split();
            for (int j = 0; j < m; j++)
                board[i, j] = int.Parse(input[j]);
        }

        List<List<bool>> vis = new List<List<bool>>();

        for (int i = 0; i < n; i++)
        {
            List<bool> row = new List<bool>();
            for (int j = 0; j < m; j++)
                row.Add(false);

            vis.Add(row);
        }

        int count = 0;
        int pMx = 0;
        Queue<(int x, int y)> q = new Queue<(int x, int y)>();        
        // 전수 검사
        for (int i = 0; i < n; i++)
        {
            for (int j = 0; j < m; j++)
            {
                if (vis[i][j] == true || board[i, j] == 0)
                    continue;

                count++;
                int mx = 0;

                q.Enqueue((i, j));
                vis[i][j] = true;

                while (q.Count > 0)
                {
                    var cur = q.Dequeue();
                    mx++;
                    for (int k = 0; k < 4; k++)
                    {
                        int nx = dx[k] + cur.y;
                        int ny = dy[k] + cur.x;

                        if (nx >= m || nx < 0 || ny >= n || ny < 0)
                            continue;

                        if (board[ny, nx] == 0 || vis[ny][nx])
                            continue;

                        q.Enqueue((ny, nx));
                        vis[ny][nx] = true;
                    }
                }
                pMx = Math.Max(pMx, mx);
            }
        }

        Console.WriteLine(count);
        Console.WriteLine(pMx);
    }
}

